mlflow/sklearn/__init__.py

```diff
@@ -455,6 +460,271 @@ def load_model(model_uri):
+          then they will be assigned different name (via appending index to the
+          eval_dataset_var_name) when autologging.
+        * storing the dataset row samples, we need log them into metric_info artifact file.
+    (4) _metric_api_call_info, it is a double level map:
+       `_metric_api_call_info[run_id][metric_name]` wil get a list of tuples, each tuple is:
+         (logged_metric_key, metric_call_command_string)
+        each call command string is like `metric_fn(arg1, arg2, ...)`
+        This data structure is used for:
+         * storing the call arguments dict for each metric call, we need log them into metric_info
+           artifact file.
+
+    Note: this class is not thread-safe.
+    Design rule for this class:
+     Because this class instance is a global instance, in order to prevent memory leak, it should
+     only holds IDs and other small objects references. This class internal data structure should
+     avoid reference to user dataset variables or model variables.
+    """
+    def __init__(self):
+        self._pred_result_id_to_dataset_name_and_run_id = {}
+        self._eval_dataset_info_map = defaultdict(lambda: defaultdict(list))
+        self._metric_api_call_info = defaultdict(lambda: defaultdict(list))
+        self._log_post_training_metrics_enabled = True
+        self._metric_info_artifact_need_update = defaultdict(lambda: False)
+
+    def should_log_post_training_metrics(self):
+        """
+        Check whether we should run patching code for autologging post training metrics.
+        This checking should surround the whole patched code due to the safe guard checking,
+        See following note.
+
+        Note: It includes checking `_SklearnTrainingSession.is_active()`, This is a safe guarding
+        for meta-estimator (e.g. GridSearchCV) case:
+          running GridSearchCV.fit, the nested `estimator.fit` will be called in parallel,
+          but, the _autolog_training_status is a global status without thread-safe lock protecting.
+          This safe guarding will prevent code run into this case.
+        """
+        return _SklearnTrainingSession.is_active() and self._log_post_training_metrics_enabled
+
+    def disable_log_post_training_metrics(self):
+
+        class LogPostTrainingMetricsDisabledScope:
+            def __enter__(inner_self):
+                inner_self.old_status = self._log_post_training_metrics_enabled
+                self._log_post_training_metrics_enabled = False
+
+            def __exit__(inner_self, exc_type, exc_val, exc_tb):
+                self._log_post_training_metrics_enabled = inner_self.old_status
+
+        return LogPostTrainingMetricsDisabledScope()
+
+    @staticmethod
+    def get_run_id_for_model(model):
+        return getattr(model, '_mlflow_run_id', None)
+
+    @staticmethod
+    def is_metrics_value_loggable(metric_value):
+        return np.isscalar(metric_value)
```

Comment:
Nit: just because it's scalar doesn't mean it can be logged as a metric:

```
>>> np.isscalar("foo")
True
```

Reply:
Yes. Shall we also filter out "string" type value ?

Reply:
Only a numeric value can be logged as a metric.

Reply:
Only allow numeric value now.